src/mlpack/methods/ann/loss_functions/KLDivergence_impl.hpp

```diff
@@ -0,0 +1,74 @@
+/**
+ * @file KLDivergence_impl.hpp
+ * @author Dakshit Agrawal
+ *
+ * Implementation of the Kullback–Leibler Divergence error function.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_ANN_LOSS_FUNCTION_KLDIVERGENCE_IMPL_HPP
+#define MLPACK_METHODS_ANN_LOSS_FUNCTION_KLDIVERGENCE_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "KLDivergence.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType>
+KLDivergence<InputDataType, OutputDataType>::KLDivergence(const bool takeMean) :
+    takeMean(takeMean)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename TargetType>
+double KLDivergence<InputDataType, OutputDataType>::Forward(
+    const InputType&& input, const TargetType&& target)
+{
+  if (takeMean)
+  {
+    return arma::as_scalar(arma::mean(
+        arma::mean(input % (arma::log(input) - arma::log(target)))));
+  }
+  else
+  {
+    return arma::accu(input % (arma::log(input) - arma::log(target)));
+  }
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename TargetType, typename OutputType>
+void KLDivergence<InputDataType, OutputDataType>::Backward(
+    const InputType&& input,
+    const TargetType&& target,
+    OutputType&& output)
+{
+  if (takeMean)
+  {
+    output = arma::mean(arma::mean(arma::log(input) - arma::log(target) + 1));
+  }
+  else
+  {
+    output = arma::accu(arma::log(input) - arma::log(target) + 1);
```

Comment:
I'm not sure this is correct, I thought it should be ``log(input/target)`` maybe I missed something?

Reply:
Ah, sorry I was talking about mean(log(input/target)), but I realized, the NLL, does also use accu, so I think we should stick to that metric.

Reply:
So, is this ready for merge?